feat: UI Option Menu and minor changes

Vlukkee opened the pull request:
To test it you need open Main Menu vs3 and play the scene.

You can only test it with a controller, test it with mouse and keyboard its very tricky.

**OPTIONS MENU**

GAME:
> FPS LIMIT - AUTO - 30 - 60 - 144
VSYNC - OFF - ON
BRIGHTNESS - 0 TO 100
RESOLUTION - AUTO - 720 - 1080 - 2K - 4K
WINDOWS MODE - FULLSCREEN - DESKTOP FULLSCREEN - BORDERLESS -  BORDER

VIDEO
> SHADOWS - ON /OFF
SSAO - ON /OFF
VSM - ON /OFF
BLOOM - ON /OFF

AUDIO // ITS NOT WORKING YET, WE NEED URI MADE THE AUDIO PART
> MASTER - 0 TO 100
MUSIC - 0 TO 100
VFX - 0 TO 100

CONTROLLER
> JUST AN IMG WITH THE CONTROLLER

MINOR CHANGES
- Change the Credits image and button.
- Create a script to show you numbers in real time.
- Modify buttons controller in the UI
- Fix an error in te play pause that make buttons dont work
- Delete old IMG that we dont use anymore

KNOW ERRORS
- Fullscreen dont work proprerly, now is in desktop fullscreen mode.

markmeliaa commented on Game/Scripts/UIOptionsMenu.h:
```diff
@@ -3,54 +3,151 @@
+
+	bool isSlider;
+	bool optionSizeLock;
+	bool resettButtonIndex;
```
```suggestion
	bool resetButtonIndex;
```

Vlukkee replied:
Changed. Thanks!

markmeliaa commented on Game/Scripts/UIOptionsMenu.cpp:
Please, check the lenght of the lines of this script, we have set in our programming rules that the column limit is 120 characters, and there are several lines in this script that are above that limit.

Vlukkee replied:
Check!

markmeliaa commented on Game/Scripts/UIOptionsMenu.cpp:
```diff
@@ -1,62 +1,658 @@
+void UIOptionsMenu::IsSizeOptionEnable()
+{
+	float4 colorSet;
+	if (actualConfig[(int)Canvas::GAME_CANVAS].options[(int)Button::WINDOWSMODE].actualOption == 0 || 
```
```suggestion
	if (actualConfig[static_cast<int>(Canvas::GAME_CANVAS)].options[(int)Button::WINDOWSMODE].actualOption == 0 || 
```

markmeliaa replied:
C style casts are not recommended in C++, `static_cast` is overall better for casting (apply this for all the script, there are several C style casts).

Vlukkee replied:
Nice to know, Tanks! Changed

markmeliaa commented on Game/Scripts/UIOptionsMenu.h:
```diff
@@ -3,54 +3,151 @@
 
 };
 
+inline void UIOptionsMenu::SetLoadFromMainMenu(bool fromMainMenu)
+{
+	loadFromMainMenu = fromMainMenu;
+}
+
+inline bool UIOptionsMenu::IsLoadFromMainMenu() const
+{
+	return loadFromMainMenu;
+}
```
We decided that we won't use `inlines` in game-related scripts, please, move those to the cpp.

Vlukkee replied:
Changed in this and others.

markmeliaa commented on Game/Scripts/UIText.cpp:
```diff
@@ -0,0 +1,87 @@
+
+		CleanInputVisuals();
+
+		for (int a=0; a < 3; a++)
```
Not needed for this pr, but for the future, try to name the variables more accordingly to what they are doing. Reading `a` as a variable gives any possible reader no information of what that it is doing, whereas that `currentComboPosition` would give the reader more information.

Also, that `3` could be turned into a `constexpr` and given a proper name (such as `comboMax`), so it would look less "hard-coded".

So your loop would look like:
```c++
for (int currentComboPosition = 0; currentComboPosition < comboMax; ++currentComboPosition)
```
And this can be extended to any variable, function or class that you create, remember to make your code readable.

P.S. Also remember that a pre-increment is preferably better in `for` loops, as we don't need to store the variable before incrementing its value (++a is better than a++).

Vlukkee replied:
Ok i dont know the pre increment for the loop. Thanks!

I set a variable to get the max number to show and change the "for" variables to get better read

```
int maxNumberToShow = inputPositions.size();

for (int currentNumToShow=0; currentNumToShow < maxNumberToShow; ++currentNumToShow)
```

markmeliaa reviewed (dismissed):
Couple typos and changes to follow the programming rules, but in general a really good job Ale, congrats!!

markmeliaa commented on Game/Scripts/UIGameManager.cpp:
```diff
@@ -16,11 +16,16 @@ REGISTERCLASS(UIGameManager);
 speedPwrUpObject(nullptr), pwrUpActive(false), savePwrUp(PowerUpType::NONE), sliderHudHealthBixFront(nullptr), 
-sliderHudHealthBixBack(nullptr), sliderHudHealthAlluraFront(nullptr), sliderHudHealthAlluraBack(nullptr)
+sliderHudHealthBixBack(nullptr), sliderHudHealthAlluraFront(nullptr), sliderHudHealthAlluraBack(nullptr),
+debbugModeObject(nullptr), imgMouse(nullptr), imgController(nullptr),inputMethod(true), prevInputMetod(nullptr)
```
`prevInputMetod` is a `bool`, not a pointer.

Vlukkee replied:
I set to true instead of false for the logic part.

Vlukkee commented on Game/Scripts/UIGameManager.h:
```diff
@@ -32,24 +32,34 @@ class UIGameManager : public Script
 
 	GameObject* mainMenuObject;
 	GameObject* hudCanvasObject;
+	GameObject* debbugModeObject;
```
OK, changed. Thanks!

Vlukkee commented on Game/Scripts/UIGameManager.h:
```diff
@@ -32,24 +32,34 @@ class UIGameManager : public Script
 	bool menuIsOpen;
 	bool pwrUpActive;
+	bool inputMethod;
+	bool prevInputMetod;
```
Cahnged. Thanks

diegusmejia117 reviewed (dismissed):
Tested it, works as intended

totisama reviewed (approved):
I got an error when I tried to change the "shadows" option to off, before that I had changed some options in the game tab. I've tried to replicate it but I couldn't get the error to show up again. Besides that everything is working nice.

![image](https://github.com/Horizons-Games/Axolotl-Engine/assets/32148906/6276d6df-a0c2-493a-a563-68465dd25312)
![image2](https://github.com/Horizons-Games/Axolotl-Engine/assets/32148906/fe89c5cc-2e5a-43d7-882c-d3cb45e4bc57)

MarcGonzalezMoratona reviewed (changes requested):
![image](https://github.com/Horizons-Games/Axolotl-Engine/assets/80526953/1b327828-7ee7-491d-b06b-4f272a6fda7a)

MarcGonzalezMoratona commented:
> ![image](https://user-images.githubusercontent.com/80526953/273568775-1b327828-7ee7-491d-b06b-4f272a6fda7a.png)

![image](https://github.com/Horizons-Games/Axolotl-Engine/assets/80526953/7fcdf3b8-1d26-4d4e-973e-44bf8e5e53c9)

MarcGonzalezMoratona commented:
OptionsConfig.txt is saving things to null and this makes the engine explode
![image](https://github.com/Horizons-Games/Axolotl-Engine/assets/80526953/af4a177e-90e2-4f76-9711-ab5fb32273bf)

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.h:
```diff
@@ -3,55 +3,141 @@
+	GAME_CANVAS = 0,
+	VIDEO_CANVAS = 1,
+	AUDIO_CANVAS = 2,
+	CONTROLL_CANVAS = 3
```
```suggestion
	CONTROLLER_CANVAS = 3
```

Vlukkee replied:
Done!

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.h:
```diff
@@ -3,55 +3,141 @@
 	void Start() override;
 	void Update(float deltaTime) override;
 
+	void ControlEnable();
```
```suggestion
	void EnableController();
```

Vlukkee replied:
I use ControllerMenuMode, because its more accurate. That not enable controls so is when we use the controller in the menu.
The contrapart will be KeyboardMenuMode.

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.h:
```diff
@@ -3,55 +3,141 @@
 	GameObject* audioOptionHover;
-	GameObject* hudOptionHover;
-	GameObject* keysOptionHover;
+	GameObject* controlOptionHover;
```
```suggestion
	GameObject* controlsOptionHover;
```

Vlukkee replied:
Done

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.h:
```diff
@@ -3,55 +3,141 @@
 	GameObject* audioOptionCanvas;
-	GameObject* hudOptionCanvas;
-	GameObject* keysOptionCanvas;
+	GameObject* controlOptionCanvas;
```
```suggestion
	GameObject* controlsOptionCanvas;
```

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.h:
```diff
@@ -3,55 +3,141 @@
 	GameObject* audioOptionButton;
-	GameObject* hudOptionButton;
-	GameObject* keysOptionButton;
+	GameObject* controlOptionButton;
```
```suggestion
	GameObject* controlsOptionButton;
```

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.h:
```diff
@@ -3,55 +3,141 @@
 	ComponentButton* audioOptionComponentButton;
-	ComponentButton* hudOptionComponentButton;
-	ComponentButton* keysOptionComponentButton;
+	ComponentButton* controlOptionComponentButton;
```
```suggestion
	ComponentButton* controlsOptionComponentButton;
```

Vlukkee replied:
done

MarcGonzalezMoratona commented on Game/Scripts/UIText.cpp:
```diff
@@ -0,0 +1,88 @@
+void UIText::FpsMetric(float time)
+{
+	currentTime = time;
+	timeDiff = currentTime - previusTime;
```
```suggestion
	timeDiff = currentTime - previousTime;
```

Vlukkee replied:
Done

MarcGonzalezMoratona commented on Game/Scripts/UIButtonControl.cpp:
```diff
@@ -10,20 +10,24 @@
-buttonHover(nullptr), isGameExit(false), isGameResume(false), setUiGameManagerObject(nullptr), UIGameManagerClass(nullptr)
+UIButtonControl::UIButtonControl() : Script(), disableObject(nullptr), enableObject(nullptr), 
+buttonComponent(nullptr), buttonHover(nullptr), isGameExit(false), isGameResume(false), 
+setUiGameManagerObject(nullptr), UIGameManagerClass(nullptr), isBackButton (false)
```
```suggestion
setUiGameManagerObject(nullptr), UIGameManagerClass(nullptr), isBackButton(false)
```

Vlukkee replied:
done

MarcGonzalezMoratona commented on Game/Scripts/UIButtonControl.cpp:
```diff
@@ -51,6 +54,12 @@ void UIButtonControl::Update(float deltaTime)
 	}
 	else if (enableObject != nullptr && disableObject != nullptr)
 	{
+		if (isBackButton != false && input->GetKey(SDL_SCANCODE_E) == KeyState::DOWN)
```
```suggestion
		if (!isBackButton && input->GetKey(SDL_SCANCODE_E) == KeyState::DOWN)
```

Vlukkee replied:
done

MarcGonzalezMoratona commented on Game/Scripts/UIGameManager.cpp:
```diff
@@ -16,11 +16,16 @@ REGISTERCLASS(UIGameManager);
 speedPwrUpObject(nullptr), pwrUpActive(false), savePwrUp(PowerUpType::NONE), sliderHudHealthBixFront(nullptr), 
-sliderHudHealthBixBack(nullptr), sliderHudHealthAlluraFront(nullptr), sliderHudHealthAlluraBack(nullptr)
+sliderHudHealthBixBack(nullptr), sliderHudHealthAlluraFront(nullptr), sliderHudHealthAlluraBack(nullptr),
+debugModeObject(nullptr), imgMouse(nullptr), imgController(nullptr),inputMethod(true), prevInputMethod(true)
```
```suggestion
debugModeObject(nullptr), imgMouse(nullptr), imgController(nullptr), inputMethod(true), prevInputMethod(true)
```

Vlukkee replied:
donep

MarcGonzalezMoratona commented on Game/Scripts/UIGameManager.cpp:
```diff
@@ -59,14 +64,41 @@ void UIGameManager::Start()
-	ModuleInput* input = App->GetModule<ModuleInput>();
+	uiTime += deltaTime;
+
+	input = App->GetModule<ModuleInput>();
```
could we do this on Start to avoid calling it every frame?

Vlukkee replied:
done

MarcGonzalezMoratona commented on Game/Scripts/UIGameManager.cpp:
```diff
@@ -59,14 +64,41 @@ void UIGameManager::Start()
 		MenuIsOpen();
 	}
 
+	// DEBBUG MODE
```
```suggestion
	// DEBUG MODE
```

Vlukkee replied:
Done!

MarcGonzalezMoratona commented on Game/Scripts/UIGameManager.cpp:
```diff
@@ -59,14 +64,41 @@ void UIGameManager::Start()
+
+	input = App->GetModule<ModuleInput>();
 
+	//IN GAME MENU
```
```suggestion
	// IN-GAME MENU
```

Vlukkee replied:
Done!

MarcGonzalezMoratona commented on Game/Scripts/UIGameManager.cpp:
```diff
@@ -229,4 +262,45 @@ void UIGameManager::SetMaxPowerUpTime(float maxPowerUpTime)
+
+void UIGameManager::InputMethodImg(bool input)
+{
+	if (currentInputTime == 0.0f && prevInputMethod != inputMethod )
```
```suggestion
	if (currentInputTime == 0.0f && prevInputMethod != inputMethod)
```

Vlukkee replied:
Done!

MarcGonzalezMoratona commented on Game/Scripts/UIGameManager.cpp:
```diff
@@ -229,4 +262,45 @@ void UIGameManager::SetMaxPowerUpTime(float maxPowerUpTime)
+	{
+		if (input)
+		{
+			prevInputMethod = input;
```
This two lines can be done outside of the if-else block and avoid code duplication
``
prevInputMethod = input;
currentInputTime++;
```

Vlukkee replied:
Done!

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.h:
```diff
@@ -3,55 +3,141 @@
 
+	void BackToLastSavedOption();
+	bool IsSlider(int header, int button, int option);
+	void IsSizeOptionEnable();
```
```suggestion
	void IsSizeOptionEnabled();
```

Vlukkee replied:
Done!

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.h:
```diff
@@ -3,55 +3,141 @@
+	void BackToLastSavedOption();
+	bool IsSlider(int header, int button, int option);
+	void IsSizeOptionEnable();
+	void IsFpsEnable();
```
```suggestion
	void IsFpsEnabled();
```

Vlukkee replied:
Done!

MarcGonzalezMoratona commented on Game/Scripts/UIText.cpp:
```diff
@@ -0,0 +1,88 @@
+
+UIText::UIText() : Script(), actualNumber(0.0f), prevNumber(0.0f)
+{
+	REGISTER_FIELD(numbersIMG, std::vector<GameObject*>);
```
or numbersImages

Vlukkee replied:
Done!

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.cpp:
```diff
@@ -1,62 +1,693 @@
+					newSelectedOption = valueSlider;
+				}
+			}
+			else // CAMBIA OPCIONES CUANDO NO ES UN SLIDER
```
```suggestion
			else // CHANGE OPTIONS WHEN IS NOT A SLIDER
```

Vlukkee replied:
Done!

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.cpp:
```diff
@@ -1,62 +1,693 @@
+	actualConfig[headerMenuPosition].options[actualButton].actualOption = newSelectedOption;
+	ApplyChanges(headerMenuPosition, actualButton, newSelectedOption);
+}
+void UIOptionsMenu::ApplyChanges(int headerMenuPosition, int actualButton, int newSelectedOption)
```
```suggestion

void UIOptionsMenu::ApplyChanges(int headerMenuPosition, int actualButton, int newSelectedOption)
```

Vlukkee replied:
Done!

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.cpp:
```diff
@@ -1,62 +1,693 @@
+		AudioOption(actualButton, newSelectedOption);
+		break;
+	case 3:
+		ControlOption();
```
```suggestion
		ControlsOption();
```

Vlukkee replied:
Done!

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.cpp:
```diff
@@ -1,62 +1,693 @@
+	optionsMenu.fromBuffer(buffer);
+	delete buffer;
+
+	for (int a = 0; a < optionsMenu.Size(); a++)
```
```suggestion
	for (int optionMenuIndex = 0; optionMenuIndex < optionsMenu.Size(); optionMenuIndex ++)
```

Vlukkee replied:
Done!

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.cpp:
```diff
@@ -1,62 +1,693 @@
+	{
+		Json canvas = optionsMenu[a];
+		CanvasOptionInfo canvasInfo;
+		for (int b = 0; b < canvas.Size(); b++)
```
```suggestion
		for (int optionButtonIndex = 0; optionButtonIndex < canvas.Size(); optionButtonIndex ++)
```

Vlukkee replied:
Done!

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.cpp:
```diff
@@ -1,62 +1,693 @@
+	CanvasOptionInfo controller;
+	actualConfig.push_back(controller);
+
+	for (int a = 0; a < actualConfig.size(); a++)
```
please use better variables names. a and b are not descriptive names of what the code is doing

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.cpp:
```diff
@@ -1,62 +1,693 @@
+
+void UIOptionsMenu::LoadDefaultOptions()
+{
+	for (int a = 0; a < buttonsAndCanvas.size()-1; a++)
```
same with these variables

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.cpp:
```diff
@@ -1,62 +1,693 @@
+
+bool UIOptionsMenu::IsSlider(int header, int button, int option)
+{
+	if (buttonsAndCanvas[header].canvas->GetChildren()[button]->GetChildren()[1]->GetChildren()[option]->
```
you can delete the if-else block and use

```suggestion
return buttonsAndCanvas[header].canvas->GetChildren()[button]->GetChildren()[1]->GetChildren()[option]->HasComponent<ComponentSlider>();
```

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.cpp:
```diff
@@ -1,62 +1,693 @@
+void UIOptionsMenu::IsSizeOptionEnable()
+{
+	float4 colorSet;
+	if (actualConfig[static_cast<int>(Canvas::GAME_CANVAS)].options[static_cast<int>(Button::WINDOWSMODE)].actualOption == 0 ||
```
you can extract this in a variable at the beginning of the method and compare it with 0 or 1 to avoid code duplication:

int actualOption = actualConfig[static_cast<int>(Canvas::GAME_CANVAS)].options[static_cast<int>(Button::WINDOWSMODE)].actualOption

Vlukkee replied:
donde

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.cpp:
```diff
@@ -1,62 +1,693 @@
+		default:
+			break;
+		}
+		if (loadFromMainMenu == false)
```
```suggestion
		if (!loadFromMainMenu)
```

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.cpp:
```diff
@@ -1,62 +1,693 @@
+		case 0:
+
+			window->SetDesktopFullscreen(true);
+			//window->SetFullscreen(true); // NOT WORKING PROPRETLY WE NEED FIX IT
```
```suggestion
			//window->SetFullscreen(true); // NOT WORKING PROPERTLY WE NEED TO FIX IT
```

Vlukkee replied:
DONE!

MarcGonzalezMoratona commented on Game/Scripts/UIOptionsMenu.cpp:
```diff
@@ -1,62 +1,693 @@
+	}
+}
+
+void UIOptionsMenu::AudioOption(int button, int option)
```
This is not finished. Are you going to address it on another PR?

Vlukkee replied:
I leave ready all for when the audio team finish add only the code necessary.

MarcGonzalezMoratona commented:
Good job Ale! 💪🏼  I found some issues and pointed out things that can be improved. Keep working on it and remember to fix conflicts! 👀

Vlukkee commented on Game/Scripts/UIGameManager.cpp:
```diff
@@ -59,14 +64,41 @@ void UIGameManager::Start()
+		}
+	}
+
+	// Player input method ture=GAMEPAD false=KEYBOARD
```
Done!

Vlukkee commented on Game/Scripts/UIOptionsMenu.h:
```diff
@@ -3,55 +3,141 @@
+	//AUDIO
+	GENERAL = 0,
+	MUSIC = 1,
+	VFX = 2,
```
Done!

Vlukkee commented on Game/Scripts/UIOptionsMenu.h:
```diff
@@ -3,55 +3,141 @@
-	GameObject* keysOptionHover;
+	GameObject* controlOptionHover;
+
+	GameObject* padTriggersIMG;
```
Done!

Vlukkee commented on Game/Scripts/UIText.cpp:
```diff
@@ -0,0 +1,88 @@
+
+void UIText::Update(float deltaTime)
+{
+	//FpsMetric(deltaTime); NEXT FEATURE WHEN SARA LET ME DO IT
```
Done

Vlukkee commented on Game/Scripts/UIText.cpp:
```diff
@@ -0,0 +1,88 @@
+
+		CleanInputVisuals();
+		
+		for (int currentNumToShow=0; currentNumToShow < maxNumberToShow; ++currentNumToShow)
```
fixed

Vlukkee commented on Game/Scripts/UIOptionsMenu.cpp:
```diff
@@ -1,62 +1,693 @@
+		default:
+			break;
+		}
+		if (loadFromMainMenu == false)
```
DONE!

Vlukkee commented on Game/Scripts/UIOptionsMenu.cpp:
```diff
@@ -1,62 +1,693 @@
+
+void UIOptionsMenu::ControlEnable()
+{
+	if (resetButtonIndex == true)
```
done

Vlukkee commented:
> OptionsConfig.txt is saving things to null and this makes the engine explode ![image](https://user-images.githubusercontent.com/80526953/273568915-af4a177e-90e2-4f76-9711-ab5fb32273bf.png)

fixed



> Some files need a name change if possible:
> 
> Remeber to fix the meta files too

Sorry but its too mutch work for now.



> > ![image](https://user-images.githubusercontent.com/80526953/273568775-1b327828-7ee7-491d-b06b-4f272a6fda7a.png)
> 
> ![image](https://user-images.githubusercontent.com/80526953/273568841-7fcdf3b8-1d26-4d4e-973e-44bf8e5e53c9.png)



> I got an error when I tried to change the "shadows" option to off, before that I had changed some options in the game tab. I've tried to replicate it but I couldn't get the error to show up again. Besides that everything is working nice.
> 
> ![image](https://user-images.githubusercontent.com/32148906/273562835-6276d6df-a0c2-493a-a563-68465dd25312.png) ![image2](https://user-images.githubusercontent.com/32148906/273562838-fe89c5cc-2e5a-43d7-882c-d3cb45e4bc57.png)

Its hard to find this but i think is fixed with the json fix

MarcGonzalezMoratona reviewed (approved):
Works well and code looks good. Nice job! thank you for fixing everything I requested!
